lisa/src/main/java/it/unive/lisa/analysis/heap/PointBasedHeap.java

```diff
@@ -0,0 +1,247 @@
+package it.unive.lisa.analysis.heap;
+
+import it.unive.lisa.analysis.BaseLattice;
+import it.unive.lisa.analysis.HeapDomain;
+import it.unive.lisa.analysis.SemanticException;
+import it.unive.lisa.program.cfg.ProgramPoint;
+import it.unive.lisa.symbolic.SymbolicExpression;
+import it.unive.lisa.symbolic.heap.AccessChild;
+import it.unive.lisa.symbolic.heap.HeapAllocation;
+import it.unive.lisa.symbolic.heap.HeapReference;
+import it.unive.lisa.symbolic.value.BinaryExpression;
+import it.unive.lisa.symbolic.value.HeapIdentifier;
+import it.unive.lisa.symbolic.value.Identifier;
+import it.unive.lisa.symbolic.value.Skip;
+import it.unive.lisa.symbolic.value.UnaryExpression;
+import it.unive.lisa.symbolic.value.ValueExpression;
+import it.unive.lisa.symbolic.value.ValueIdentifier;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.collections.CollectionUtils;
+
+/**
+ * A point-based heap implementation that abstracts heap locations depending on
+ * their allocation sites, namely the position of the code wehre heap locations
+ * are generated. All heap locations that are generated at the same allocation
+ * sites are abstracted into a single unique identifier.
+ * 
+ * @author <a href="mailto:vincenzo.arceri@unive.it">Vincenzo Arceri</a>
+ */
+public class PointBasedHeap extends BaseLattice<PointBasedHeap> implements HeapDomain<PointBasedHeap> {
+
+	private static final PointBasedHeap TOP = new PointBasedHeap();
+
+	private static final PointBasedHeap BOTTOM = new PointBasedHeap();
+
+	private final Collection<ValueExpression> rewritten;
+
+	private final Map<Identifier, Collection<ValueExpression>> allocationSites;
+
+	/**
+	 * Builds a new instance of PointBasedHeap, with an unique rewritten
+	 * expression {@link Skip}.
+	 */
+	public PointBasedHeap() {
+		this(new Skip());
+	}
+
+	private PointBasedHeap(ValueExpression rewritten) {
+		this(Collections.singleton(rewritten), new HashMap<>());
+	}
+
+	private PointBasedHeap(Collection<ValueExpression> rewritten,
+			Map<Identifier, Collection<ValueExpression>> allocationSites) {
+		this.rewritten = rewritten;
+		this.allocationSites = allocationSites;
+	}
+
+	@Override
+	public PointBasedHeap smallStepSemantics(SymbolicExpression expression, ProgramPoint pp) throws SemanticException {
+
+		if (expression instanceof AccessChild) {
+			PointBasedHeap container = smallStepSemantics((((AccessChild) expression).getContainer()), pp);
+			return new PointBasedHeap(container.getRewrittenExpressions(), allocationSites);
```

Comment:
A couple of points here:
* we are not using the `allocationSites` from `container`? Is it right to discard the newly-created sites from the evaluation of the container?
* If the previous point is right, we don't need to create a new instance here: we can return the unmodified `container` (or the object resulting from the evaluation of the child -- see above note)